lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -440,28 +456,34 @@ LogicalResult static applyNoBlackBoxStyleDataTaps(const AnnoPathValue &target,
                << "\n"
                << i->getParentOfType<FModuleOp>().getNameAttr() << ">"
                << i.getNameAttr(););
-    auto srcModule =
-        dyn_cast<FModuleOp>(srcTarget->ref.getModule().getOperation());
-    Value refSendBase;
-    ImplicitLocOpBuilder refSendBuilder(srcModule.getLoc(), srcModule);
-    // Set the insertion point for the RefSend, it should be dominated by the
-    // srcTarget value. If srcTarget is a port, then insert the RefSend
-    // at the beggining of the module, else define the RefSend at the end of
-    // the block that contains the srcTarget Op.
-    if (srcTarget->ref.getImpl().isOp()) {
-      refSendBase = srcTarget->ref.getImpl().getOp()->getResult(0);
-      refSendBuilder.setInsertionPointAfter(srcTarget->ref.getOp());
-    } else if (srcTarget->ref.getImpl().isPort()) {
-      refSendBase = srcModule.getArgument(srcTarget->ref.getImpl().getPortNo());
-      refSendBuilder.setInsertionPointToStart(srcModule.getBodyBlock());
+    // The RefSend value can be either generated by the instance of an external
+    // module or a RefSendOp.
+    if (!sendVal) {
```

Comment:
This `sendVal` is mutable state that is defined and set very far away. Is there a way to tighten this up?

Reply:
I agree, it's too far away, but I couldn't tighten it up further. Only option is to move much of the code to its own function.